upgrade_tests/upgrade_compact_storage.py

```diff
@@ -178,3 +178,50 @@ def test_upgrade_with_dropped_compact_storage_index(self):
                      [{'col1': '50', 'column1': None, 'pk': '5', 'value': None}])
         assert (list(session.execute("SELECT * FROM ks.compact_table WHERE pk = '5'")) ==
                      [{'col1': '50', 'column1': None, 'pk': '5', 'value': None}])
+
+    def test_downgrade_after_failed_upgrade(self):
+        self.prepare(cassandra_version=VERSION_311)
+        node = self.cluster.nodelist()[0]
+        session = self.patient_cql_connection(node, row_factory=dict_factory)
+
+        session.execute("CREATE KEYSPACE ks WITH replication = {'class': 'SimpleStrategy','replication_factor': '1' };")
+        session.execute("CREATE TABLE ks.compact_table (pk ascii PRIMARY KEY, col1 ascii) WITH COMPACT STORAGE")
+        session.execute("CREATE INDEX ON ks.compact_table(col1)")
+
+        for i in range(1, 10):
+            session.execute("INSERT INTO ks.compact_table (pk, col1) VALUES ('{pk}', '{col1}')".format(pk=i, col1=i * 10))
+
+        assert (list(session.execute("SELECT * FROM ks.compact_table WHERE col1 = '50'")) ==
+                     [{'pk': '5', 'col1': '50'}])
+        assert (list(session.execute("SELECT * FROM ks.compact_table WHERE pk = '5'")) ==
+                     [{'pk': '5', 'col1': '50'}])
+
+        logging.debug("Upgrading to current version")
+
+        self.fixture_dtest_setup.allow_log_errors = True
+
+        node.stop(wait_other_notice=False)
+        node.set_install_dir(version=VERSION_TRUNK)
+        try:
+            node.start(wait_other_notice=False, wait_for_binary_proto=False, verbose=False)
+        except NodeError:
+            print("error")  # ignore
+        time.sleep(5)
+        # After restart, it won't start
+        errors = len(node.grep_log("Compact Tables are not allowed in Cassandra starting with 4.0 version"))
+        assert errors > 0
+
+        logging.debug("Downgrading to 3.11")
+        node.set_install_dir(version=VERSION_311)
+        logger.debug("Set new cassandra dir for %s: %s" % (node.name, node.get_install_dir()))
+        self.cluster.set_install_dir(version=VERSION_311)
+        self.fixture_dtest_setup.reinitialize_cluster_for_different_version()
+        node.start(wait_other_notice=False, wait_for_binary_proto=False)
+        logger.debug('Starting %s on new version (%s)' % (node.name, VERSION_311))
+
+        session = self.patient_cql_connection(node, row_factory=dict_factory)
+
+        assert (list(session.execute("SELECT * FROM ks.compact_table WHERE col1 = '50'")) ==
+                [{'pk': '5', 'col1': '50'}])
+        assert (list(session.execute("SELECT * FROM ks.compact_table WHERE pk = '5'")) ==
+                [{'pk': '5', 'col1': '50'}])
```

Comment:
After checking that the data is still there after downgrading, we could continue the test dropping `COMPACT STORAGE` and upgrading to 4.0 again. This would have some overlapping with `test_upgrade_with_dropped_compact_storage`, but I think that it wouldn't be exactly the same case and it would give us a good reproduction of the entire use case. WDYT?

Reply:
Done